R/derive_vars_dt_dtm_utils.R

```diff
@@ -260,3 +262,515 @@ get_partialdatetime <- function(dtc) {
+#'
+#' @keywords internal
+#'
+#' @importFrom lubridate ymd_hms ymd rollback
+adjust_last_day_imputation <- function(imputed_dtc, partial, is_datetime) {
+  if_else(
+    is.na(partial[["day"]]),
+    if (is_datetime) {
+      strftime(
+        rollback(ymd_hms(imputed_dtc) + months(1)),
+        format = "%Y-%m-%dT%H:%M:%S",
+        tz = "UTC"
+      )
+    } else {
+      strftime(
+        rollback(ymd(imputed_dtc) + months(1)),
+        format = "%Y-%m-%d"
+      )
+    },
+    imputed_dtc
+  )
+}
+
+
+#' Impute Missing Values
+#'
+#' @description
+#' `r lifecycle::badge("stable")`
+#'
+#' This is a helping function for `get_dt_dtm_range()`
+#' Imputes missing values in partial date/time components using target values.
+#'
+#' @param partial A list of partial date/time components.
+#' @param target A list of target values for imputation.
+#' @param components A character vector of component names.
+#'
+#' @returns A list of imputed date/time components.
+#'
+#' @keywords internal
+impute_values <- function(partial, target, components) {
+  imputed <- vector("list", length(components))
+  names(imputed) <- components
+  for (c in components) {
+    imputed[[c]] <- if_else(is.na(partial[[c]]), target[[c]], partial[[c]])
+  }
+  imputed
+}
+
+#' Format Imputed Date/Datetime
+#'
+#' @description
+#' `r lifecycle::badge("stable")`
+#'
+#' This is a helping function for `get_dt_dtm_range()`
+#' Formats imputed date/datetime components into a string representation.
+#'
+#' @param imputed A list of imputed date/time components.
+#' @param is_datetime A logical indicating whether it's a datetime format.
+#'
+#' @returns A character vector of formatted date/datetime strings.
+#'
+#' @details
+#' The function formats the imputed components into "YYYY-MM-DD" for dates
+#' and "YYYY-MM-DDTHH:MM:SS" for datetimes. It replaces any string containing
+#' 'x' with NA.
+#'
+#' @keywords internal
+format_imputed_dtc <- function(imputed, is_datetime) {
+  if (is_datetime) {
+    dtc <- paste0(
+      paste(imputed[["year"]], imputed[["month"]], imputed[["day"]], sep = "-"),
+      "T",
+      paste(imputed[["hour"]], imputed[["minute"]], imputed[["second"]], sep = ":")
+    )
+  } else {
+    dtc <- paste(imputed[["year"]], imputed[["month"]], imputed[["day"]], sep = "-")
+  }
+  if_else(str_detect(dtc, "x"), NA_character_, dtc)
+}
+#' Propagate NA Values
+#'
+#' @description
+#' `r lifecycle::badge("stable")`
+#'
+#' This is a helping function for `get_dt_dtm_range()`.
+#' Propagates NA values through date/time components.
+#'
+#' @param partial A list of partial date/time components.
+#'
+#' @returns A list of date/time components with propagated NA values.
+#'
+#' @details
+#' This function ensures that if a higher-order component (e.g., month) is NA,
+#' all lower-order components (e.g., day, hour, etc.) are also set to NA.
+#'
+#' @keywords internal
+propagate_na_values <- function(partial) {
+  for (i in 2:6) {
```

Comment:
```suggestion
  for (i in 2:length(partial) {
```
Then it can be used for both dates and datetimes.